pineappl_cli/src/export/applgrid.rs

```diff
@@ -18,32 +19,61 @@ fn reconstruct_subgrid_params(grid: &Grid, order: usize, bin: usize) -> Result<S
         .subgrids()
         .slice(s![order, bin, ..])
         .iter()
-        .map(|subgrid| {
+        .flat_map(|subgrid| {
             subgrid
                 .mu2_grid()
-                .iter()
+                .into_iter()
                 .map(|&Mu2 { ren, fac }| {
                     if !approx_eq!(f64, ren, fac, ulps = 128) {
                         bail!("subgrid has mur2 != muf2, which APPLgrid does not support");
                     }
 
                     Ok(fac)
                 })
-                .collect::<Result<Vec<_>>>()
+                .collect_vec()
         })
-        .collect::<Result<_>>()?;
-    let mut mu2_grid: Vec<_> = mu2_grid.into_iter().flatten().collect();
-    mu2_grid.dedup_by(|a, b| approx_eq!(f64, *a, *b, ulps = 128));
-    let mu2_grid = mu2_grid.as_slice();
-
-    if let &[fac] = mu2_grid {
-        result.set_q2_bins(1);
-        result.set_q2_max(fac);
-        result.set_q2_min(fac);
+        .collect::<Result<Vec<_>>>()?
+        .into_iter()
+        .sorted_unstable_by(|a, b| a.total_cmp(b))
+        .dedup_by(|&a, &b| approx_eq!(f64, a, b, ulps = 128))
+        .collect_vec();
+
+    let x_grid = grid
+        .subgrids()
+        .slice(s![order, bin, ..])
```

Comment:
This probably doesn't work quite as expected. APPLgrid stores different channels in a single `appl_igrid`, which has common parameter values `NQ2`, `NQ2min`, `NQ2max`, `NQorder`, `Nx`, `xmin`, `xmax`, `xorder`. This means that:
- if several channels have different `SubgridParam`/`ExtraSubgridParam` values we should error out, unless this is a DIS grid in which we ignore the second dimension. In PineAPPL the second dimension can be either `x1` or `x2`
- the `flat_map` should therefore probably be a simple `map` and we must make sure that for all channels the entries are the same
- the same applies also for `mu_grid`

Reply:
The `OPTIMIZE_SUBGRID_TYPE` optimization can change the ends of the `x` and `mu2` grid points, right? So I would check if the grid points are equal modulo some endpoints, and then use the union as the `x` points that will be passed to APPLgrid?

Reply:
This would address one point of it, yes.